server/storage/endpoint/gc_key_space.go

```
@@ -0,0 +1,189 @@
+type KeySpaceGCSafePoint struct {
+	SpaceID   string `json:"space_id"`
+	SafePoint uint64 `json:"safe_point,omitempty"`
+}
+
+// KeySpaceGCSafePointStorage defines the storage operations on KeySpaces' safe points
+type KeySpaceGCSafePointStorage interface {
+	// Service safe point interfaces.
+	SaveServiceSafePoint(spaceID string, ssp *ServiceSafePoint) error
+	LoadServiceSafePoint(spaceID, serviceID string) (*ServiceSafePoint, error)
+	LoadMinServiceSafePoint(spaceID string, now time.Time) (*ServiceSafePoint, error)
+	RemoveServiceSafePoint(spaceID, serviceID string) error
+	// GC safe point interfaces.
+	SaveKeySpaceGCSafePoint(spaceID string, safePoint uint64) error
+	LoadKeySpaceGCSafePoint(spaceID string) (uint64, error)
+	LoadAllKeySpaceGCSafePoints(withGCSafePoint bool) ([]*KeySpaceGCSafePoint, error)
+}
+
+var _ KeySpaceGCSafePointStorage = (*StorageEndpoint)(nil)
+
+// SaveServiceSafePoint saves service safe point under given key-space.
+func (se *StorageEndpoint) SaveServiceSafePoint(spaceID string, ssp *ServiceSafePoint) error {
+	if ssp.ServiceID == "" {
+		return errors.New("service id of service safepoint cannot be empty")
+	}
+	key := KeySpaceServiceSafePointPath(spaceID, ssp.ServiceID)
+	value, err := json.Marshal(ssp)
+	if err != nil {
+		return err
+	}
+	return se.Save(key, string(value))
+}
+
+// LoadServiceSafePoint reads ServiceSafePoint for the given key-space ID and service name.
+// Return nil if no safepoint exist for given service or just expired.
+func (se *StorageEndpoint) LoadServiceSafePoint(spaceID, serviceID string) (*ServiceSafePoint, error) {
+	key := KeySpaceServiceSafePointPath(spaceID, serviceID)
+	value, err := se.Load(key)
+	if err != nil || value == "" {
+		return nil, err
+	}
+	ssp := &ServiceSafePoint{}
+	if err := json.Unmarshal([]byte(value), ssp); err != nil {
+		return nil, err
+	}
+	if ssp.ExpiredAt < time.Now().Unix() {
+		go func() {
+			se.Remove(key)
```

What if the removal failed?

---

Currently we are just ignoring the error and attempt to delete next time. (same as the `APIv1` behavior )https://github.com/tikv/pd/blob/562586ceebae1dc41cf7c04e9770558d9ea5a067/server/storage/endpoint/gc_safe_point.go#L102

And since we are removing them asynchronously, the best we can do is to put something in log like `failed to remove key {key}, err: {err}`. 

Having few expired safePoints does not really interfere with min calculation, it's only a problem if they accumulate. One solution is to have something check if there's too many expired safePoints once per few updates, but I'm not sure if that's really necessary, considering we skipped it in v1.

---

Can we log it firstly?

---

Sure